mlflow/utils/databricks_utils.py

```diff
@@ -738,54 +738,107 @@ def _init_databricks_cli_config_provider(entry_point):
+                    fallback_api_token_option = notebook_utils.getContext().apiToken()
+                    logger.logUsage(
+                        "refreshableTokenNotFound",
+                        {"api_url": api_url},
+                        None,
+                    )
+                    if fallback_api_token_option.isDefined():
+                        api_token = fallback_api_token_option.get()
 
-            ssl_trust_all = entry_point.getDriverConf().workflowSslTrustAll()
+                ssl_trust_all = entry_point.getDriverConf().workflowSslTrustAll()
 
-            # Fallback to the default behavior if we were unable to find a token.
-            if api_token is None or api_url is None:
-                return DefaultConfigProvider().get_config()
+                # Fallback to the default behavior if we were unable to find a token.
+                if api_token is None or api_url is None:
+                    return DefaultConfigProvider().get_config()
 
-            return DatabricksConfig.from_token(
-                host=api_url, token=api_token, insecure=ssl_trust_all
-            )
+                return DatabricksConfig.from_token(
+                    host=api_url, token=api_token, insecure=ssl_trust_all
+                )
+    elif dbr_major_minor_version >= (10, 3):
+        class DynamicConfigProvider(DatabricksConfigProvider):
+            def get_config(self):
+                try:
+                    from dbruntime.databricks_repl_context import get_context
+
+                    ctx = get_context()
+                    if ctx and ctx.apiUrl and ctx.apiToken:
+                        return DatabricksConfig.from_token(
+                            host=ctx.apiUrl, token=ctx.apiToken, insecure=ctx.sslTrustAll)
+                except Exception as e:
+                    print(
+                        "Unexpected internal error while constructing `DatabricksConfig` "
+                        "from REPL context: {e}".format(e=e),
+                        file=stderr,
+                    )
+                # Invoking getContext() will attempt to find the credentials related to the
+                # current command execution, so it's critical that we execute it on every
+                # get_config().
+                api_token_option = notebook_utils.getContext().apiToken()
+                api_url_option = notebook_utils.getContext().apiUrl()
+                ssl_trust_all = entry_point.getDriverConf().workflowSslTrustAll()
+
+                # Fallback to the default behavior if we were unable to find a token.
+                if not api_token_option.isDefined() or not api_url_option.isDefined():
+                    return DefaultConfigProvider().get_config()
+
+                return DatabricksConfig.from_token(
+                    host=api_url_option.get(), token=api_token_option.get(), insecure=ssl_trust_all
+                )
+    else:
+        class DynamicConfigProvider(DatabricksConfigProvider):
```

Comment:
Ref: spark/driver/python/python_shell/dbruntime/databricksCli.py in universe 9.1 <= version <= 10.2